### PR TITLE
feat(connector): [Klarna] Add support for Capture, Psync, Refunds and Rsync flows

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Add support for Capture, Psync, Refunds and Rsync for Klarna Connector

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create a Merchant 
```
{
  "merchant_id": "merchant_{{$timestamp}}",
  "locker_id": "m0010",
  "merchant_name": "NewAge Retailer",
  "merchant_details": {
    "primary_contact_person": "John Test",
    "primary_email": "JohnTest@test.com",
    "primary_phone": "sunt laborum",
    "secondary_contact_person": "John Test2",
    "secondary_email": "JohnTest2@test.com",
    "secondary_phone": "cillum do dolor id",
    "website": "https://www.example.com",
    "about_business": "Online Retail with a wide selection of organic products for North America",
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US"
    }
  },
  "return_url": "https://google.com/success",
  "webhook_details": {
    "webhook_version": "1.0.1",
    "webhook_username": "ekart_retail",
    "webhook_password": "password_ekart@123",
    "payment_created_enabled": true,
    "payment_succeeded_enabled": true,
    "payment_failed_enabled": true
  },
  "sub_merchants_enabled": false,
  "metadata": {
    "city": "NY",
    "unit": "245"
  },
  "primary_business_details": [
    {
      "country": "US",
      "business": "food"
    }
  ]
}
```
___________________________
2. Create API Key 
```
{
  "name": "API Key 1",
  "description": null,
  "expiration": "2038-01-19T03:14:08.000Z"
}
```
_________________________________
3. Create MCA
```
{
    "connector_type": "fiz_operations",
    "connector_name": "klarna",
    "connector_account_details": {
        "auth_type": "BodyKey",
        "api_key": "Password",
        "key1": "username"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "invoke_sdk_client",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "klarna"
                }
            ]
        }
    ],
    "metadata": {
        "city": "NY",
        "unit": "245",
        "klarna_region": "NorthAmerica"
    }
}
```
_____________________________________
In metadata  try 3 values and do a Klarna Payment via SDK for Klarna Connector for each `klarna_region`
 i.  
 ```
"metadata": {
        "city": "NY",
        "unit": "245",
        "klarna_region": "Europe"
    }
```
  ii.
 ```
"metadata": {
        "city": "NY",
        "unit": "245",
        "klarna_region": "NorthAmerica"
    }
```
iii.
 ```
"metadata": {
        "city": "NY",
        "unit": "245",
        "region_based_endpoint": "Oceania"
    }
```
  _______________________________________________________________
  Create a payment via sdk with capture_method = `manual`
```
{
      currency: "USD",
      amount: 5000,
      order_details: [
        {
          product_name: "Apple iphone 15",
          quantity: 1,
          amount: 5000,
        },
      ],
      confirm: false,
      capture_method: "manual",
      authentication_type: "three_ds",
      customer_id: "hyperswitch_sdk_demo_id",
      email: "hyperswitch_sdk_demo_id@gmail.com",
      description: "Hello this is description",
      connector_metadata: {
        noon: {
          order_category: "applepay",
        },
      },
      metadata: {
        udf1: "value1",
        new_customer: "true",
        login_date: "2019-09-10T10:11:12Z",
      },
      billing: {
        address: {
          line1: "1467",
          line2: "Harrison Street",
          line3: "Harrison Street",
          city: "San Fransico",
          state: "California",
          zip: "94122",
          country: "US",
          first_name: "joseph",
          last_name: "Doe",
        },
        phone: {
          number: "8056594427",
          country_code: "+91",
        },
      },
      }
```
____________________________________________
Capture the Payment
```
curl --location 'http://localhost:8080/payments/{payment_id}/capture' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: ******************' \
--data '{
  "amount_to_capture": 5000,
  "statement_descriptor_name": "Joseph",
  "statement_descriptor_suffix": "JS"
}'
```

Response:
```
{
    "payment_id": "pay_bH1TZc5hf9jqEbFSz7GH",
    "merchant_id": "merchant_1716967519",
    "status": "succeeded",
    "amount": 5000,
    "net_amount": 5000,
    "amount_capturable": 0,
    "amount_received": 5000,
    "connector": "klarna",
    "client_secret": "pay_bH1TZc5hf9jqEbFSz7GH_secret_4nmkltwKTUENqgGmjgsi",
    "created": "2024-05-29T11:30:39.409Z",
    "currency": "USD",
    "customer_id": "hyperswitch_sdk_demo_id",
    "customer": {
        "id": "hyperswitch_sdk_demo_id",
        "name": null,
        "email": "hyperswitch_sdk_demo_id@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Hello this is description",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "pay_later",
    "payment_method_data": {
        "pay_later": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 5000,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "requires_shipping": null
        }
    ],
    "email": "hyperswitch_sdk_demo_id@gmail.com",
    "name": null,
    "phone": null,
    "return_url": "http://localhost:9060/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "klarna",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "3a20817c-8956-4ef7-9627-12967b4e9e17",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": {
        "apple_pay": null,
        "airwallex": null,
        "noon": {
            "order_category": "applepay"
        }
    },
    "feature_metadata": null,
    "reference_id": "3a20817c-8956-4ef7-9627-12967b4e9e17",
    "payment_link": null,
    "profile_id": "pro_W1dQQFAsWElgN0Uk9yjx",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_CR12bLCzbHKYSPJtvqzK",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-29T11:45:39.409Z",
    "fingerprint": null,
    "browser_info": {
        "language": "en-GB",
        "time_zone": -330,
        "ip_address": "::1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1728,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 1117,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-29T11:31:21.792Z",
    "charges": null,
    "frm_metadata": null
}
```
______________________________________
update payment_attempt set status = 'pending' and update payment_intent set status = 'processing'

Do a psync 
```
curl --location 'http://localhost:8080/payments/{payment_id}?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: **********************************\'
```
Response:
```
{
    "payment_id": "pay_NeIO8Rjzc10dYJrYsYz3",
    "merchant_id": "merchant_1716967519",
    "status": "succeeded",
    "amount": 5000,
    "net_amount": 5000,
    "amount_capturable": 0,
    "amount_received": 5000,
    "connector": "klarna",
    "client_secret": "pay_NeIO8Rjzc10dYJrYsYz3_secret_SyaVWvLioqrpMlUZIh9t",
    "created": "2024-05-29T09:55:50.071Z",
    "currency": "USD",
    "customer_id": "hyperswitch_sdk_demo_id",
    "customer": {
        "id": "hyperswitch_sdk_demo_id",
        "name": null,
        "email": "hyperswitch_sdk_demo_id@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Hello this is description",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "pay_later",
    "payment_method_data": {
        "pay_later": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 5000,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "requires_shipping": null
        }
    ],
    "email": "hyperswitch_sdk_demo_id@gmail.com",
    "name": null,
    "phone": null,
    "return_url": "http://localhost:9060/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "klarna",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "af406563-993d-48b8-83a4-56114856d7f9",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": {
        "apple_pay": null,
        "airwallex": null,
        "noon": {
            "order_category": "applepay"
        }
    },
    "feature_metadata": null,
    "reference_id": "af406563-993d-48b8-83a4-56114856d7f9",
    "payment_link": null,
    "profile_id": "pro_W1dQQFAsWElgN0Uk9yjx",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_CR12bLCzbHKYSPJtvqzK",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-29T10:10:50.071Z",
    "fingerprint": null,
    "browser_info": {
        "language": "en-GB",
        "time_zone": -330,
        "ip_address": "::1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 2560,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 1440,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-29T09:58:44.297Z",
    "charges": null,
    "frm_metadata": null
}
```
______________________________________

Create a Refund
```
curl --location 'http://localhost:8080/refunds' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: ***************' \
--data '{
  "payment_id": "{payment_id}",
  "amount": 5000,
  "reason": "Customer returned product",
  "refund_type": "instant",
  "metadata": {
    "udf1": "value1",
    "new_customer": "true",
    "login_date": "2019-09-10T10:11:12Z"
  }
}'
```
Response
```
{
    "refund_id": "ref_uzvvTjY9GNEEuiKtTgEa",
    "payment_id": "pay_bH1TZc5hf9jqEbFSz7GH",
    "amount": 5000,
    "currency": "USD",
    "status": "pending",
    "reason": "Customer returned product",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "error_message": null,
    "error_code": null,
    "created_at": "2024-05-29T11:41:48.050Z",
    "updated_at": "2024-05-29T11:41:48.050Z",
    "connector": "klarna",
    "profile_id": "pro_W1dQQFAsWElgN0Uk9yjx",
    "merchant_connector_id": "mca_CR12bLCzbHKYSPJtvqzK",
    "charges": null
}
```
______________________________________________________________
Do Rsync
```
curl --location 'http://localhost:8080/refunds/{refund_id}' \
--header 'Accept: application/json' \
--header 'api-key: *************'
```
Response:
```
{
    "refund_id": "ref_uzvvTjY9GNEEuiKtTgEa",
    "payment_id": "pay_bH1TZc5hf9jqEbFSz7GH",
    "amount": 5000,
    "currency": "USD",
    "status": "succeeded",
    "reason": "Customer returned product",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "error_message": null,
    "error_code": null,
    "created_at": "2024-05-29T11:41:48.050Z",
    "updated_at": "2024-05-29T11:41:48.050Z",
    "connector": "klarna",
    "profile_id": "pro_W1dQQFAsWElgN0Uk9yjx",
    "merchant_connector_id": "mca_CR12bLCzbHKYSPJtvqzK",
    "charges": null
}
```

___________________________
Create a payment with capture_method = `manual`
and then Void the payment. 
```
curl --location 'http://localhost:8080/payments/{paymnet_id}/cancel' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: *****************' \
--data '{
  "cancellation_reason": "requested_by_customer"
}'
```
Response
```
{
    "payment_id": "pay_xqVbolfYyBBpqaZPBaDB",
    "merchant_id": "merchant_1716967519",
    "status": "cancelled",
    "amount": 5000,
    "net_amount": 5000,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": "klarna",
    "client_secret": "pay_xqVbolfYyBBpqaZPBaDB_secret_3A4OQikb4oz99tyDb4UT",
    "created": "2024-05-29T10:52:04.202Z",
    "currency": "USD",
    "customer_id": "hyperswitch_sdk_demo_id",
    "customer": {
        "id": "hyperswitch_sdk_demo_id",
        "name": null,
        "email": "hyperswitch_sdk_demo_id@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Hello this is description",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "pay_later",
    "payment_method_data": {
        "pay_later": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 5000,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "requires_shipping": null
        }
    ],
    "email": "hyperswitch_sdk_demo_id@gmail.com",
    "name": null,
    "phone": null,
    "return_url": "http://localhost:9060/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": "requested_by_customer",
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "klarna",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "e8922d04-7f80-49fe-a7a8-e40f5c49014f",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": {
        "apple_pay": null,
        "airwallex": null,
        "noon": {
            "order_category": "applepay"
        }
    },
    "feature_metadata": null,
    "reference_id": "e8922d04-7f80-49fe-a7a8-e40f5c49014f",
    "payment_link": null,
    "profile_id": "pro_W1dQQFAsWElgN0Uk9yjx",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_CR12bLCzbHKYSPJtvqzK",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-29T11:07:04.202Z",
    "fingerprint": null,
    "browser_info": {
        "language": "en-GB",
        "time_zone": -330,
        "ip_address": "::1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 2560,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 1440,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-29T10:52:36.717Z",
    "charges": null,
    "frm_metadata": null
}
```

  Create a payment via sdk with capture_method = `automatic`
```
{
      currency: "USD",
      amount: 5000,
      order_details: [
        {
          product_name: "Apple iphone 15",
          quantity: 1,
          amount: 5000,
        },
      ],
      confirm: false,
      capture_method: "automatic",
      authentication_type: "three_ds",
      customer_id: "hyperswitch_sdk_demo_id",
      email: "hyperswitch_sdk_demo_id@gmail.com",
      description: "Hello this is description",
      connector_metadata: {
        noon: {
          order_category: "applepay",
        },
      },
      metadata: {
        udf1: "value1",
        new_customer: "true",
        login_date: "2019-09-10T10:11:12Z",
      },
      billing: {
        address: {
          line1: "1467",
          line2: "Harrison Street",
          line3: "Harrison Street",
          city: "San Fransico",
          state: "California",
          zip: "94122",
          country: "US",
          first_name: "joseph",
          last_name: "Doe",
        },
        phone: {
          number: "8056594427",
          country_code: "+91",
        },
      },
      }
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/klarna.rs [hunk 1]
```diff
@@ -297,14 +311,179 @@ impl
...
+
+    fn get_request_body(
+        &self,
+        req: &types::PaymentsCaptureRouterData,
+        _connectors: &settings::Connectors,
+    ) -> CustomResult<RequestContent, errors::ConnectorError> {
+        let connector_router_data = klarna::KlarnaRouterData::try_from((
+            &self.get_currency_unit(),
+            req.request.currency,
+            req.request.amount_to_capture,
```
Comment (Narayanbhat166) [inline crates/router/src/connector/klarna.rs:349]: we are already passing req, then why do we need to pass these two fields

File: crates/router/src/connector/klarna.rs [hunk 2]
```diff
@@ -297,14 +311,179 @@ impl
...
+    fn handle_response(
+        &self,
+        data: &types::PaymentsCaptureRouterData,
+        event_builder: Option<&mut ConnectorEvent>,
+        res: Response,
+    ) -> CustomResult<types::PaymentsCaptureRouterData, errors::ConnectorError> {
+        match res.headers {
+            Some(headers) => {
+                let capture_id = connector_utils::get_http_header("Capture-Id", &headers)
+                    .attach_printable("Missing refund id in headers")
```
Comment (Narayanbhat166) [inline crates/router/src/connector/klarna.rs:384]:
```suggestion
                    .attach_printable("Missing capture id in headers")
```

File: crates/router/src/connector/klarna.rs [hunk 3]
```diff
@@ -297,14 +311,179 @@ impl
...
+            .connector_transaction_id
+            .get_connector_transaction_id()
+            .change_context(errors::ConnectorError::RequestEncodingFailed)?;
+        let endpoint =
+            build_region_specific_endpoint(self.base_url(connectors), &req.connector_meta_data)?;
+
+        Ok(format!(
+            "{}{}{}",
+            endpoint, "ordermanagement/v1/orders/", order_id,
+        ))
```
Comment (Narayanbhat166) [inline crates/router/src/connector/klarna.rs:441]: can you do the similar change here?

**Review by SamraatBansal — DISMISSED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

File: crates/router/src/connector/klarna.rs (hunk 3 above)
Comment (swangi-kumari) [reply crates/router/src/connector/klarna.rs:441]: will take this up in next pr

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**